### DISCUSSION TITLE
Performing BulkCopy for IDataReader

### DISCUSSION THREAD

**Discussion by pavrosa** [linq2db/linq2db#4000] [Q&A]
I would like to feed BulkCopy from an IDataReader instead of an IEnumerable<T>, but I can't find an easy way to do it. The SqlBulkCopy class (System.Data.SqlClient / Microsoft.Data.SqlClient) provides the required functionality for SQL Server, but is platform dependent:-(
Can anyone help?

**Comment by MaceWindu** [comment]
What are you trying to achieve? linq2db already use `IDataReader` for bulk copy under the hood

**Reply by pavrosa** [reply]
The thing is, I have an IDataReader and I'd like to insert its data into a table via BulkCopy, but that method expects an IEnumerable&lt;T&gt; while I don't have any corresponding entity class, and I'd rather not have such.
I need this specifically for PostgreSQL and I've already tried looking under the hood, but PostgreSQLBulkCopy doesn't seem to offer such an option.

**Reply by MaceWindu** [reply]
Yep, npgsql implements their own bulk copy API, which doesn't accept data reader https://github.com/npgsql/npgsql/blob/main/src/Npgsql/NpgsqlBinaryImporter.cs#L375
